images/hub/jupyterhub_config.py

```diff
@@ -177,6 +178,25 @@ def get_config(key, default=None):
     c.DummyAuthenticator.password = get_config('auth.dummy.password', None)
 elif auth_type == 'tmp':
     c.JupyterHub.authenticator_class = 'tmpauthenticator.TmpAuthenticator'
+elif auth_type == 'custom':
+    # full_class_name looks like "myauthenticator.MyAuthenticator".
+    # To create a docker image with this class availabe, you can just have the
+    # following Dockerifle:
+    #   FROM jupyterhub/k8s-hub:v0.4
+    #   RUN pip3 install myauthenticator
+    full_class_name = get_config('auth.custom.class_name')
+    c.JupyterHub.authenticator_class = full_class_name
+
+    (module, class_name) = full_class_name.split('.')
+    the_module = importlib.import_module(module)
+    klass=getattr(the_module, class_name)
+
+    class_params = get_config('auth.custom.params')
```

Comment:
With traitlets config, you don't want to be setting anything on the class or importing it. You only want to touch the config object section with the appropriate name, e.g.

```python
c.ClassName.trait = 'foo'
```

or the dict equivalent:

```python
c['ClassName'].update({'trait': 'foo'})
```

If you have custom config in a dict, it could look like this:

```yaml
auth:
  type: "custom"
  custom:
    className: "hashauthenticator.HashAuthenticator"
    traits:
      secret_key: "834e6cace4747e966c002737191619f583be814557d3f11e358232a9a28dfa7d"
```

```python
# jupyterhub_config.py
full_class_name = get_config('auth.custom.class_name')
c.JupyterHub.authenticator_class = full_class_name
auth_class_name = full_class_name.rsplit('.', 1)[-1]
auth_config = c[auth_class_name]
auth_config.update(get_config('auth.custom.traits') or {})
```

Alternately, and perhaps better, the helm yaml could have a config section that is just a passthrough for `jupyterhub_config.json`:

```yaml
jupyterhubExtraConfig:
  ClassName:
    trait: value
```

for which the helm config would have the responsibility of dumping yaml to json.